### DISCUSSION TITLE
Replacing custom route validations with Symfony conditions

### DISCUSSION THREAD

**Discussion by ollieread** [laravel/framework#33749] [General]
I've recently encountered issue #31774 which talks about Laravels custom route validators not working for cached routes. In there is a comment by @driesvints pointing out that since Laravel uses Symfonys latest router these can't be honoured.

Symfony does support the adding of conditions, referred to as [expressions](https://symfony.com/doc/current/routing.html#matching-expressions) to routes, allowing you to conditional match routes based on header values and the likes.

This uses [Symfonys ExpressionLanguage](https://symfony.com/doc/current/components/expression_language/syntax.html) so it could support that, or a nicer syntax, something like:

```php
Route::get('/a-route')
    ->name('route')
    ->conditions([
        Conditions::header('Content-Type')->is('application/json+v3')
    ]);
```

Could be useful for versioned APIs.

```php
Route::group(function() {
    Route::get('/', [V1UserController::class, 'index'])->conditions([
        Conditions::header('Content-Type')->is('application/json+v1')
    ]);
    Route::get('/', [V2UserController::class, 'index'])->conditions([
        Conditions::header('Content-Type')->is('application/json+v2')
    ]);
    Route::get('/', [V3UserController::class, 'index'])->conditions([
        Conditions::header('Content-Type')->is('application/json+v3')
    ]);
});
```

I think this is something that Laravel should implement, and I'd be happy to do so and make a PR. @driesvints recommended that I start a discussion in here and see what people think.

Is this something people would be interested in?

**Comment by driesvints** [comment]
Is there anything preventing you from using middleware instead or is it just about a nicer syntax? Does this work well with compiled routes & route caching?

**Reply by ollieread** [reply]
I'm currently looking into how it handles compiled and cached routes, but I can't find anything indicating that there would be an issue with those.

As for the middleware issue, the problem is that you can't really conditionally prevent a route from loading with it. In the example code above there are three matches for the URI, but the conditions change which is dispatched. At least that is my understanding, I'm still looking into this but the documentation definitely implies that.

**Reply by driesvints** [reply]
Gotcha. I think a PR for this could be considered but ultimately it's up to Taylor. This could be interesting maybe.
